chainntnfs/btcdnotify/btcd.go

```diff
@@ -406,59 +437,496 @@ out:
+			}
+		}
 	}
-	return &txConf, nil
+
+	return errTxNotFound
+}
+
+// rescanChainForOutPoint rescans the node's chain starting from the given
+// block hash while watching for an already spent outpoint. The rescan will
+// dispatch the spend notification.
+func (b *BtcdNotifier) rescanChainForOutPoint(blockHash *chainhash.Hash,
+	outpoint *wire.OutPoint) error {
+
+	ops := []*wire.OutPoint{outpoint}
+	if err := b.chainConn.Rescan(blockHash, nil, ops); err != nil {
+		chainntnfs.Log.Errorf("Rescan for spend notification txout "+
+			"failed: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// spendDetailsFromPeers looks up the spending transaction of the outpoint by
+// fetching blocks from the node's peers and manually scanning them for a
+// transaction input that spends the outpoint. The blocks fetched will range
+// from the height hint to the current best height. The height hint should
+// indicate the height where the spent outpoint was included in. If the spending
+// transaction is found, it dispatches the spend notification for the outpoint.
+func (b *BtcdNotifier) spendDetailsFromPeers(outpoint *wire.OutPoint,
+	heightHint, bestHeight uint32) error {
+
+	btcNet, err := b.chainConn.GetCurrentNet()
+	if err != nil {
+		return err
+	}
+
+	// Set up a buffered channel to retrieve the messages from our peers.
+	msgChan := make(chan interface{}, 1)
+
+	// Set up the necessary message listeners for our peers. In this case,
+	// we only care about receiving `block` and `notfound` messages as a
+	// response to `getdata` messages.
+	listeners := peer.MessageListeners{
+		OnBlock: func(_ *peer.Peer, msg *wire.MsgBlock, buf []byte) {
+			block := btcutil.NewBlockFromBlockAndBytes(msg, buf)
+			msgChan <- block
+		},
+		OnNotFound: func(_ *peer.Peer, msg *wire.MsgNotFound) {
+			msgChan <- msg
+		},
+	}
+
+	// Keep track of the last height to avoid re-scanning from the start
+	// when trying the next peer.
+	lastHeight := heightHint
+
+	peerInfo, err := b.chainConn.GetPeerInfo()
+	if err != nil {
+		return fmt.Errorf("unable to fetch peer info: %v", err)
+	}
+
+	for _, info := range peerInfo {
+		// Parse the peer's supported services.
+		serviceFlags, err := parsePeerServices(info.Services)
+		if err != nil {
+			continue
+		}
+
+		// Make sure the peer we're attempting to connect to is a full
+		// node and supports witness data.
+		if serviceFlags&wire.SFNodeNetwork != wire.SFNodeNetwork ||
+			serviceFlags&wire.SFNodeWitness != wire.SFNodeWitness {
+			continue
+		}
+
+		// Attempt to connect to the peer. If we are not successful,
+		// continue to the next peer.
+		p, err := connectToOutboundPeer(info.Addr, netToParams[btcNet],
+			listeners, msgChan)
+		if err != nil {
+			continue
+		}
+
+		// Begin looking for the spending transaction by fetching blocks
+		// from our peer at every height and scanning them.
+		for height := lastHeight; height <= bestHeight; height++ {
+			blockHash, err := b.chainConn.GetBlockHash(int64(height))
```

Comment:
Not sure how easy it will be, but could we extract common code by putting the `b.chainConn` behind an interface that could be used by both `btcd` and `bitcoind``